example-crates/example_esplora/src/main.rs

```diff
@@ -0,0 +1,341 @@
+                }));
+            }
+
+            // drop lock on graph and chain
+            drop((graph, chain));
+
+            (
+                client
+                    .update_tx_graph_without_keychain(
+                        spks,
+                        txids,
+                        outpoints,
+                        scan_options.parallel_requests,
+                    )
+                    .context("syncing transaction updates")?,
+                Default::default(),
+            )
+        }
+    };
+
+    println!();
+
+    let (heights_to_fetch, tip) = {
+        let chain = &*chain.lock().unwrap();
+        let heights_to_fetch = update_graph.missing_heights(chain).collect::<Vec<_>>();
+        let tip = chain.tip();
+        (heights_to_fetch, tip)
+    };
+
+    #[cfg(debug_assertions)]
+    println!(
+        "old chain: {:?}",
+        tip.iter()
+            .flat_map(CheckPoint::iter)
+            .map(|cp| cp.height())
+            .collect::<Vec<_>>()
+    );
+    println!("prev tip: {}", tip.as_ref().map_or(0, CheckPoint::height));
+    println!("missing blocks: {:?}", heights_to_fetch);
+
+    let update = client
+        .update_local_chain(tip, heights_to_fetch)
+        .context("scanning for blocks")?;
+
+    #[cfg(debug_assertions)]
+    println!(
+        "new chain: {:?}",
+        update.tip.iter().map(|cp| cp.height()).collect::<Vec<_>>()
+    );
+    println!("new tip: {}", update.tip.height());
+
+    // check that all anchors are part of the new tip's history
+    #[cfg(debug_assertions)]
+    {
+        use bdk_chain::bitcoin::BlockHash;
+        use bdk_chain::collections::HashMap;
+        let chain_heights = update
+            .tip
+            .iter()
+            .map(|cp| (cp.height(), cp.hash()))
+            .collect::<HashMap<u32, BlockHash>>();
+        for (anchor, _) in update_graph.all_anchors() {
+            assert_eq!(anchor.anchor_block.height, anchor.confirmation_height);
+            assert!(chain_heights.contains_key(&anchor.anchor_block.height));
+
+            let remote_hash = chain_heights
+                .get(&anchor.confirmation_height)
+                .expect("must have block");
+
+            // inform about mismatched blocks
+            if remote_hash != &anchor.anchor_block.hash {
+                println!("mismatched block @ {}!", anchor.confirmation_height);
+                println!("\t- anchor_block: {}", anchor.anchor_block.hash);
+                println!("\t-   from_chain: {}", remote_hash);
+            }
+        }
+    }
+
+    let db_changeset: LocalChangeSet<Keychain, ConfirmationTimeAnchor> = {
+        let mut chain = chain.lock().unwrap();
+        let mut graph = graph.lock().unwrap();
+
+        let chain_changeset = chain.apply_update(update)?;
+
+        let indexed_additions = {
+            let mut additions = IndexedAdditions::default();
+            let (_, index_additions) = graph.index.reveal_to_target_multi(&update_keychain_indices);
```

Comment:
Can we redesign this so `reveal_to_target_multi` is only done in the `scan` branch and not for `sync`. We really want to get across that `scan` is the unusual choice which is slow and requires this patch up logic. Most of the time the application should be picky about what data it fetches by looking at what is likely to have updates.. We also need to a comment explaining why reveal_to_target_multi **must** be done first when you're using a stop gap type scan.